contracts/external/traders/LimitOrders.sol

```diff
@@ -0,0 +1,583 @@
+    )
+        external
+    {
+        approveOrderInternal(msg.sender, orderHash);
+    }
+
+    // ============ Only-Solo Functions ============
+
+    /**
+     * Allows traders to make trades approved by this smart contract. The active trader's account is
+     * the takerAccount and the passive account (for which this contract approves trades
+     * on-behalf-of) is the makerAccount.
+     *
+     * @param  inputMarketId   The market for which the trader specified the original amount
+     * @param  outputMarketId  The market for which the trader wants the resulting amount specified
+     * @param  makerAccount    The account for which this contract is making trades
+     * @param  takerAccount    The account requesting the trade
+     * @param  inputWei        The change in token amount for the makerAccount for the inputMarketId
+     * @param  data            Arbitrary data passed in by the trader
+     * @return                 The AssetAmount for the makerAccount for the outputMarketId
+     */
+    function getTradeCost(
+        uint256 inputMarketId,
+        uint256 outputMarketId,
+        Account.Info memory makerAccount,
+        Account.Info memory takerAccount,
+        Types.Par memory /* oldInputPar */,
+        Types.Par memory /* newInputPar */,
+        Types.Wei memory inputWei,
+        bytes memory data
+    )
+        public
+        onlySolo(msg.sender)
+        returns (Types.AssetAmount memory)
+    {
+        Order memory order = parseOrder(data);
+
+        validateOrder(
+            order,
+            data,
+            makerAccount,
+            takerAccount
+        );
+
+        return getTradeCostInternal(
+            inputMarketId,
+            outputMarketId,
+            inputWei,
+            order
+        );
+    }
+
+    /**
+     * Allows users to send this contract arbitrary data.
+     *
+     * @param  accountInfo  The account from which the data is being sent
+     * @param  data         Arbitrary data given by the sender
+     */
+    function callFunction(
+        address /* sender */,
+        Account.Info memory accountInfo,
+        bytes memory data
+    )
+        public
+        onlySolo(msg.sender)
+    {
+        CallFunctionData memory cfd = parseCallFunctionData(data);
+        if (cfd.callType == CallFunctionType.Approve) {
+            approveOrderInternal(accountInfo.owner, cfd.orderHash);
+        } else {
+            assert(cfd.callType == CallFunctionType.Cancel);
+            cancelOrderInternal(accountInfo.owner, cfd.orderHash);
+        }
+    }
+
+    // ============ Private Storage Functions ============
+
+    /**
+     * Cancels an order as long as it is not already canceled.
+     */
+    function cancelOrderInternal(
+        address canceler,
+        bytes32 orderHash
+    )
+        private
+    {
+        Require.that(
+            g_status[canceler][orderHash] != OrderStatus.Canceled,
+            FILE,
+            "Cannot cancel canceled order"
```

Comment:
Is it better to revert rather than do nothing...? not sure

Reply:
I have it emit the log again and keep the state. It should be fine that way